File: cpp/leetcode/040.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        if(candidates.empty()) return {};
        vector<vector<int>> res;
        vector<int> ans;
        sort(candidates.begin(), candidates.end());
        help(0, target, candidates, ans, res);
        return res;
    }
    
    void help(int start, int target, vector<int>& candidates, vector<int>& ans, vector<vector<int>> &res){
        if(target == 0){
            res.push_back(ans);
            return;
        }
        for(int i = start; i < candidates.size() && candidates[i] <= target; i++){
            if(i > start && candidates[i] == candidates[i-1]) continue;
            ans.push_back(candidates[i]);
            help(i+1, target - candidates[i],candidates, ans, res);
            ans.pop_back();
        }
    }
};

File: python/leetcode/040.py
class Solution(object):
    def combinationSum2(self, nums, target):
        nums.sort()
        res = []
        self.dfs(0, target, nums, [], res)
        return res

    def dfs(self, start, target, nums, ans, res):
        if target == 0:
            res.append(ans)
            return
        for i in range(start, len(nums)):
            if target < nums[i]: return
            if i == start or nums[i] != nums[i-1]:
                self.dfs(i+1, target - nums[i], nums, ans + [nums[i]], res)